src/main/java/com/vmware/vipclient/i18n/messages/service/ProductService.java

```diff
@@ -71,28 +75,81 @@ public ProductService(BaseDTO dto) {
-        while((locales == null || locales.isEmpty()) && msgSourceQueueIter.hasNext()){
-            DataSourceEnum dataSource = msgSourceQueueIter.next();
-            ProductOpt opt = dataSource.createProductOpt(dto);
-            for (String languageTag : opt.getSupportedLocales()) {
-                locales.add(Locale.forLanguageTag(languageTag));
-            }
-            // If failed to get locales from the data source, log the error.
-            if (locales == null || locales.isEmpty()) {
-                logger.error(ConstantsMsg.GET_LOCALES_FAILED, dataSource.toString());
-            }
+        Set<String> supportedLangTags = new HashSet<>();
+        while(msgSourceQueueIter.hasNext() && supportedLangTags.isEmpty()) {
+            supportedLangTags = getSupportedLanguageTags(msgSourceQueueIter.next());
+        }
+        return supportedLangTags;
+    }
+
+    public Set<String> getSupportedLanguageTags(DataSourceEnum dataSource) {
+        CacheService cs = new CacheService(new MessagesDTO(dto));
+        MessageCacheItem cacheItem = cs.getCacheOfLocales(dataSource);
+        if (cacheItem != null) {
+            if (cacheItem.isExpired())
+                refreshCacheItemTask(cacheItem, dataSource);
+            return cacheItem.getCachedData().keySet();
+        } else {
+            cacheItem = createCacheItem(dataSource);
+            if (cacheItem == null)
+                return new HashSet<>();
+            return cacheItem.getCachedData().keySet();
         }
-        return locales;
     }
 
     public boolean isSupportedLocale(Locale locale) {
         return getSupportedLocales().contains(LocaleUtility.fmtToMappedLocale(locale));
     }
 
+    private void refreshCacheItem(final MessageCacheItem cacheItem, DataSourceEnum dataSource) {
+        long timestampOld = cacheItem.getTimestamp();
+        dataSource.createProductOpt(dto).getSupportedLocales(cacheItem);
+        long timestamp = cacheItem.getTimestamp();
+        if (timestampOld == timestamp) {
+            logger.debug(FormatUtils.format(ConstantsMsg.GET_LOCALES_FAILED, dataSource.toString()));
+        }
+    }
+
+    private void refreshCacheItemTask(MessageCacheItem cacheItem, DataSourceEnum dataSource) {
+        Callable<MessageCacheItem> callable = () -> {
+            try {
+                refreshCacheItem(cacheItem, dataSource);
+                return cacheItem;
+            } catch (Exception e) {
+                return null;
+            }
+        };
+        FutureTask<MessageCacheItem> task = new FutureTask<>(callable);
+        Thread thread = new Thread(task);
+        thread.start();
+    }
+
+    private MessageCacheItem createCacheItem (DataSourceEnum dataSource) {
+        CacheService cs = new CacheService(new MessagesDTO(dto));
+        MessageCacheItem cacheItem = new MessageCacheItem();
+        refreshCacheItem(cacheItem, dataSource);
+        if (!cacheItem.getCachedData().isEmpty()) {
```

Comment:
I think this judgement is not needed. No matter the locale list is empty or not, it should be cached, or it will always try to fetch locale list from data source if last fetch fails, which may impact the performance.